File: src/levels/L16-combos-con-extras/ts/extras_decorator.ts
// region: WithOatMilk
class WithOatMilk extends Product {
  constructor(private inner: Product) { super() }
  make() { const drink = this.inner.make(); drink.extras.push('avena'); return drink }
}
// endregion

// region: WithShot
class WithShot extends Product {
  constructor(private inner: Product) { super() }
  make() { const drink = this.inner.make(); drink.extras.push('shot'); return drink }
}
// endregion

class CoffeeMenu {
  static EXTRAS: Record<string, new (drink: Product) => Product> = { avena: WithOatMilk, shot: WithShot }

  // El café del combo se envuelve con cada extra: sigue siendo un Product más dentro del combo.
  // region: CoffeeMenu#drink_for
  drinkFor(extras: string[]) { return extras.reduce<Product>((drink, e) => new CoffeeMenu.EXTRAS[e](drink), new Latte()) }
  // endregion
}

File: src/levels/L16-combos-con-extras/extras_decorator.rb
require "delegate"
# region: WithOatMilk
class WithOatMilk < SimpleDelegator
  def make = super.tap { |parts| parts.last.extras << :avena }
end
# endregion

# region: WithShot
class WithShot < SimpleDelegator
  def make = super.tap { |parts| parts.last.extras << :shot }
end
# endregion

class CoffeeMenu
  EXTRAS = { avena: WithOatMilk, shot: WithShot }

  # El café del combo se envuelve con cada extra: sigue siendo un Product más dentro del combo.
  # region: CoffeeMenu#drink_for
  def drink_for(extras) = extras.reduce(Latte.new) { |drink, e| EXTRAS.fetch(e).new(drink) }
  # endregion
end
